core/eolearn/core/eodata_io.py

```diff
@@ -29,41 +29,84 @@ def save_eopatch(eopatch, filesystem, patch_location, features=..., overwrite_pe
+
+    elif sys_is_windows() and overwrite_permission is OverwritePermission.OVERWRITE_FEATURES:
+        _check_letter_case_collisions(eopatch_features, fs_features)
 
-    if overwrite_permission is OverwritePermission.ADD_ONLY or \
-            (sys_is_windows() and overwrite_permission is OverwritePermission.OVERWRITE_FEATURES):
-        fs_features = list(walk_filesystem(filesystem, patch_location))
     else:
-        fs_features = []
+        _check_letter_case_collisions(eopatch_features, [])
 
-    _check_case_matching(eopatch_features, fs_features)
+    features_to_save = []
+    for ftype, fname, path in eopatch_features:
+        feature_io = FeatureIO(filesystem, path)
+        data = eopatch[(ftype, fname)]
+        file_format = FileFormat.NPY if ftype.is_raster() else FileFormat.PICKLE
 
-    if overwrite_permission is OverwritePermission.ADD_ONLY:
-        _check_add_only_permission(eopatch_features, fs_features)
+        features_to_save.append((feature_io, data, file_format, compress_level))
 
-    ftype_folder_map = {(ftype, fs.path.dirname(path)) for ftype, _, path in eopatch_features if not ftype.is_meta()}
-    for ftype, folder in ftype_folder_map:
+    # Cannot be done before due to lazy loading (this would delete the files before the data is loaded)
+    if overwrite_permission is OverwritePermission.OVERWRITE_PATCH and patch_exists:
+        filesystem.removetree(patch_location)
+        if patch_location != '/':  # avoid redundant filesystem.makedirs if the location is '/'
+            filesystem.makedirs(patch_location, recreate=True)
+
+    ftype_folders = {fs.path.dirname(path) for ftype, _, path in eopatch_features if not ftype.is_meta()}
+    for folder in ftype_folders:
         if not filesystem.exists(folder):
             filesystem.makedirs(folder, recreate=True)
 
-    features_to_save = ((FeatureIO(filesystem, path),
-                         eopatch[(ftype, fname)],
-                         FileFormat.NPY if ftype.is_raster() else FileFormat.PICKLE,
-                         compress_level) for ftype, fname, path in eopatch_features)
-
     with concurrent.futures.ThreadPoolExecutor() as executor:
         # The following is intentionally wrapped in a list in order to get back potential exceptions
         list(executor.map(lambda params: params[0].save(*params[1:]), features_to_save))
 
+    if overwrite_permission is not OverwritePermission.OVERWRITE_PATCH:
+        remove_redundant_files(filesystem, eopatch_features, fs_features, compress_level)
+
+
+def remove_redundant_files(filesystem, eopatch_features, filesystem_features, current_compress_level):
+    """ Removes files that should have been overwriten but were not due to different compression levels
+    """
+    files_to_remove = []
+    saved_features = {(ftype, fname) for ftype, fname, _ in eopatch_features}
+    for ftype, fname, path in filesystem_features:
+        if fname is ... and not ftype.is_meta():
+            continue
```

Comment:
I'm  quite sure this case cannot happen because of the last if-sentence in `walk_filesystem` function.

Reply:
will check and update accordingly